src/main/java/threads/task/Main.java

```diff
@@ -0,0 +1,171 @@
+        Supplier<Map<String, List<String>>> messages = () -> Client.messages
+                .entrySet()
+                .stream()
+                .filter(map -> !map.getKey().equals(getCurrentThreadName()))
+                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+        return messages.get();
+    }
+
+    public void readNewMessages() {
+        System.out.println(getCurrentThreadName() + " unread messages: " + getMessagesFromOtherThreads().toString());
+        printMessagesNewMessagesFromOtherThreads();
+        clearHistory();
+    }
+
+    private void printMessagesNewMessagesFromOtherThreads() {
+        getMessagesFromOtherThreads()
+                .forEach((key, value) -> System.out.println("From: " + key + " message: " +
+                        value
+                                .stream()
+                                .collect(Collector.of(
+                                        StringBuilder::new,
+                                        (sb, message) -> sb.append(message).append(", "),
+                                        StringBuilder::append,
+                                        sb -> {
+                                            var messages = sb.toString();
+                                            if (messages.endsWith(", "))
+                                                return messages.substring(0, messages.length() - 2);
+                                            return messages;
+                                        }
+                                ))
+                ));
+    }
+
+    private void printMessagesOfInterruptedThreads() {
+        System.out.println("From: " + getCurrentThreadName() + " message: " + messages
+                .get(getCurrentThreadName())
+                .stream()
+                .collect(Collector.of(
+                        StringBuilder::new,
+                        (sb, message) -> sb.append(message).append(", "),
+                        StringBuilder::append,
+                        sb -> {
+                            var messages = sb.toString();
+                            if (messages.endsWith(", "))
+                                return messages.substring(0, messages.length() - 2);
+                            return messages;
+                        }
+                )) + " interrupt ______________________________"
+        );
+    }
+
+    public void isThreadActive() {
+        var isOneOfThreadWroteMoreThan3Messages = getMessagesFromOtherThreads()
+                .values()
+                .stream()
+                .anyMatch(x -> x.size() >= 3);
+        var isCurrentThreadWroteMoreThan3Messages = messages.get(getCurrentThreadName()).size() >= 3;
+        if (isOneOfThreadWroteMoreThan3Messages || isCurrentThreadWroteMoreThan3Messages) {
+            printMessagesOfInterruptedThreads();
+            clearOwnHistory();
+            currentThread().interrupt();
```

Comment:
here we should (probably) interrupt not current thread, but those, which already sent max number of messages

Reply:
Done